### DISCUSSION TITLE
Use as framework - Relations queries with custom model

### DISCUSSION THREAD

**Discussion by mxmaxime** [pocketbase/pocketbase#1637] [Q&A]
Hi,
First of all, thanks for this project!
I'm trying it out to see if it can fit my needs. I need to use it as a framework, and I need to directly make sql queries for a huge part of my app.

I can't find how to make relationships working, even for basic cases, nothing on the dbx library neither pocketbase.

Example, I have a simple relation: a `device` has one `location`.
I tried things like this:

```go
// My locations model
type Locations struct {
    models.BaseModel
    Structure string `db:"structure" json:"structure"`
}

func (m *Locations) TableName() string {
    return "locations" // the name of your collection
}

// Devices model
type Devices struct {
    models.BaseModel
    DeviceId string `db:"device_id" json:"device_id"`
    Name string `db:"name" json:"name"`
    Type string `db:"type" json:"type"`
    LocationId string `db:"locations_id" json:"locations_id"`
    Location *locations.Locations
}

func (m *Devices) TableName() string {
    return "devices" // the name of your collection
}

// And the query:
func (q *DevicesQuery) Device(with_location bool) (*Devices, error) {
    device := &Devices{}

    err := q.DevicesQuery().
        LeftJoin("locations", dbx.NewExp("locations.id = locations_id")).
        One(device)
    if err != nil {
        return nil, err
    }

    return device, nil
}
```

It provides the good sql, but in "locations" I have a nil pointer. Tried multiple things, without pointer ect...
```
[0.00ms] SELECT {{devices}}.* FROM `devices` LEFT JOIN `locations` ON locations.id = locations_id
```
seems ok, I tried the query and it worked.
```
sqlite> select * from devices LEFT JOIN locations ON locations.id = locations_id LIMIT 1;
2023-01-18 18:00:32.464Z|98798709|Raspberry pi principal|t964x15gw21pg99|raspberry pi 4|2023-01-18 18:00:32.464Z|wnmeprrjysuvtvi|2023-01-18 17:59:57.069Z|wnmeprrjysuvtvi|maison|sal à manger|2023-01-18 17:59:57.069Z
```

Any hint?

**Comment by ganigeorgiev** [comment]
If you want to populate the relation record(s) as part of the query, you'll have to add the relation fields to the `SELECT` statement, but the caveat is that each rel field will have to be prefixed with the field name (`structRelField.column`) so that it can be populated. Or in other words, your query will have to be something like:

```go
err := q.DevicesQuery().
	AndSelect("locations.id as `location.id`, locations.structure as `location.structure`") // fill only the id and structure
	LeftJoin("locations", dbx.NewExp("locations.id = locations_id")).
	One(device)
```

**Reply by ganigeorgiev** [reply]
This is the example from the [dbx documentation](https://github.com/pocketbase/dbx#executing-queries):


```go
type User struct {
	id     int
	Type   int    `db:"-"`
	MyName string `db:"name"`
	Profile
	Address Address `db:"addr"`
}

type Profile struct {
	Age int
}

type Address struct {
	City string
}
```

* `User.id`: not populated because the field is not exported;
* `User.Type`: not populated because the `db` tag is `-`;
* `User.MyName`: to be populated from the `name` column, according to the `db` tag;
* `Profile.Age`: to be populated from the `age` column, since `Profile` is an anonymous field;
* `Address.City`: to be populated from the `addr.city` column, since `Address` is a named field of struct type
  and its fields will be prefixed with `addr.` according to the `db` tag.

_(your use case is similar to the `Address.City`)_

**Reply by mxmaxime** [reply]
It works, thanks for your quick answer and explanations!
It's a bit verbose but it's not an issue, I'll create a func helper if needed later on.

**Reply by ganigeorgiev** [reply]
Alternatively, you can also make the `Location` a method of the `Device` struct and resolve it lazily (aka. send a separate query just for the Location model and optionally cache it in some private field to prevent requesting it second time). SQLite is extremely performant for small queries and you usually don't have to worry about the n+1 issue even if you have to iterate multiple `Device` structs. You can also check https://www.sqlite.org/np1queryprob.html.
